src/controller.js

```diff
@@ -0,0 +1,601 @@
+  }
+
+  /**
+   * Obtain the ETag for the committed multipart upload
+   */
+  __getMultipartEtag(doc, bucket, key) {
+    return this.__getResponseProperty(doc, 'CompleteMultipartUploadResult', 'ETag', bucket, key);
+  }
+
+  // Return a list of 2-tuple's that are header key and value pairings for the
+  // headers to specify as the ACL for an object
+  __determinePermissionsHeaders(permissions) {
+    permissions = runSchema(permissions, schemas.permissions.required());
+    let {acl, read, write, readAcp, writeAcp, fullControl} = permissions;
+    if (acl) {
+      return [['x-amz-acl', permissions.acl]];
+    }
+
+    let perms = [];
+
+    if (read) {
+      perms.push(['x-amz-grant-read', read]);
+    }
+
+    if (write) {
+      perms.push(['x-amz-grant-write', write]);
+    }
+
+    if (readAcp) {
+      perms.push(['x-amz-grant-read-acp', readAcp]);
+    }
+
+    if (writeAcp) {
+      perms.push(['x-amz-grant-write-acp', writeAcp]);
+    }
+
+    if (fullControl) {
+      perms.push(['x-amz-grant-full-control', fullControl]);
+    }
+
+    return perms;
+  }
+
+  __s3hostname(bucket) {
+    let hostname = `${bucket}.${this.s3host}`;
+    if (this.s3port) {
+      hostname += ':' + this.s3port;
+    }
+    return hostname;
+  }
+ 
+  /**
+   * Initiate a Multipart upload and return the UploadIp that
+   * Amazon has assigned for this multipart upload
+   * http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadInitiate.html
+   */
+  async initiateMultipartUpload(opts) {
+
+    opts = runSchema(opts, Joi.object().keys({
+      bucket: schemas.bucket.required(),
+      key: schemas.key.required(),
+      sha256: schemas.sha256.required(),
+      size: schemas.mpSize.required(),
+      permissions: schemas.permissions,
+    }).optionalKeys('permissions'));
+
+    let {bucket, key, sha256, size, permissions} = opts;
+
+    if (size <= 0) {
+      // Each part must have a non-zero size
+      throw new Error('Objects must be more than 0 bytes');
+    } else if (size > 5 * 1024 * 1024 * 1024 * 1024) {
+      // The entire file must be lower than 5 TB
+      throw new Error('Object must total fewer than 5 TB'); 
+    }
+    let unsignedRequest = {
+      service: 's3',
+      region: this.region,
+      method: 'POST',
+      protocol: this.s3protocol,
+      hostname: this.__s3hostname(bucket),
+      path: `/${key}?uploads=`,
+      headers: {
+        'x-amz-meta-content-sha256': sha256,
+        'x-amz-meta-content-length': size,
+      }
+    };
+
+    // If we have permissions, set those values on the headers
+    if (permissions) {
+      let permHeaders = this.__determinePermissionsHeaders(permissions);
+      for (let tuple of permHeaders) {
+        unsignedRequest.headers[tuple[0]] = tuple[1];
+      }
+    }
+
+    let signedRequest = aws4.sign(unsignedRequest);
+
+    let response = await this.runner({
+      req: this.__serializeRequest(signedRequest)
```

Comment:
It's not complicated at all.  Once you've implemented the code to generate a multipart part upload request and the code to run said request, it's actually trivially easy to call all methods the same way.  It also ensures that all of the signing code that will eventually be used for the workers works 100%.

I also think that the `aws-sdk` library is painful to work with.  This adds another dependency for no benefit.

Which test case is not covered that you feel should be?